libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -434,9 +473,52 @@ traceProp ::
   (MockChainState era -> MockChainState era -> prop) ->
   Gen prop
 traceProp proof numTxInTrace gsize f = do
-  trace1 <- genTrace proof numTxInTrace gsize makeSomePools
+  trace1 <- genTrace proof numTxInTrace gsize (initStableFields proof)
   pure (f (_traceInitState trace1) (lastState trace1))
 
+forEachEpochTrace ::
+  forall era prop.
+  ( Testable prop,
+    Reflect era
+  ) =>
+  Proof era ->
+  Int ->
+  GenSize ->
+  (Trace (MOCKCHAIN era) -> prop) ->
+  Gen Property
+forEachEpochTrace proof tracelen genSize f = do
+  let newEpoch tr1 tr2 = nesEL (mcsNes tr1) /= nesEL (mcsNes tr2)
+  trc <- case proof of
+    Babbage _ -> genTrace proof tracelen genSize (initStableFields proof)
+    Alonzo _ -> genTrace proof tracelen genSize (initStableFields proof)
+    Allegra _ -> genTrace proof tracelen genSize (initStableFields proof)
+    Mary _ -> genTrace proof tracelen genSize (initStableFields proof)
+    Shelley _ -> genTrace proof tracelen genSize (initStableFields proof)
+  let splits = splitTrace newEpoch trc
+  return . conjoin $ f <$> init splits
```

Comment:
Using `init` here is terrible. If `splits` turns out to be empty all we will get in stderr is:
```
*** Exception: Prelude.init: empty list
``` 

Without any clue where it came from. Can I ask you please to rewrite it somehow to avoid this "function that should be banned"?

It also would be useful to know at least an index of the trace or anything else that identifies which of the elements of `conjoin` has failed. Eg. `conjoin [counterexample ("Trace index failed " ++ show ix) p | (ix, p) <- zip [0..] properties]`

Reply:
done